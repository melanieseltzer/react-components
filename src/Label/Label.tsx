import React, { ReactNode } from 'react';
import styled from 'styled-components';

export interface Label extends React.InputHTMLAttributes<HTMLLabelElement> {
  border?: number;
  color?: string;
}

interface StyledProps {
  float?: boolean;
}

const Label = (props: Label): ReactNode => {
  const { children } = props;
  return <StyledLabel {...props}>{children}</StyledLabel>;
};

export default Label;

const StyledLabel = styled.label`
  font-size: 1rem;
  transition: all 0.2s;
  color: ${({ color }) => (color ? color : '#222')};
  /* Leverage props when used in context of floating label */
  ${({ float }: StyledProps) =>
    float &&
    `
      position: absolute;
      top: 0;
      left: 0;
      &:hover {
        cursor: text;
      }
    `}
`;
