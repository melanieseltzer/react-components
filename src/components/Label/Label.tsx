import React from 'react';
import styled from 'styled-components';

export interface Props extends React.InputHTMLAttributes<HTMLLabelElement> {
  color?: string;
  htmlFor?: string;
  width?: string;
  /** float prop is passed from Floating and not meant to be used explicitely */
  float?: boolean;
  /** Pass it any class to extend styling */
  className?: string;
}

const Label = (props: Props) => {
  const { color = '#aaa', children, className } = props;

  return (
    <LabelWrapper color={color} className={className} {...props}>
      {children}
    </LabelWrapper>
  );
};

const LabelWrapper = styled.label`
  transition: all 0.2s;
  color: ${(props: Props) => props.color};
  /* Leverage props when used in context of floating label */
  ${(props: Props) =>
    props.float &&
    `
      position: absolute;
      top: 0;
      left: 0;
      &:hover {
        cursor: text;
      }
    `}
`;

export default Label;
