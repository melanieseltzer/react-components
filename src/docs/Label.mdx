---
name: Label
menu: Components
route: /label
---

import { Playground, Props } from 'docz';
import Input from '../components/Input/Input';
import Label from '../components/Label/Label';
import Flex from '../components/Flex/Flex';

# Label

An HTML label component. Should be used with [Input](/input).

## Props

<Props of={Label} />

## Basic usage

Label above input:

<Playground>
  <Flex direction="column">
    <Flex.Item>
      <Label className="custom-class" htmlFor="name2">
        Name
      </Label>
    </Flex.Item>
    <Flex.Item>
      <Input
        className="custom-class"
        id="name2"
        name="name2"
        placeholder="John Smith"
        border={1}
        width="100%"
      />
    </Flex.Item>
  </Flex>
</Playground>

Label next to input:

<Playground>
  <Flex justifyContent="space-around" alignItems="center">
    <Flex.Item flex="1 1 5%">
      <Label className="custom-class" htmlFor="name2" color="tomato">
        Name
      </Label>
    </Flex.Item>
    <Flex.Item flex="1 1 90%">
      <Input
        className="custom-class"
        id="name2"
        name="name2"
        placeholder="John Smith"
        color="tomato"
        border={1}
        borderBottom
        width="100%"
      />
    </Flex.Item>
  </Flex>
</Playground>
