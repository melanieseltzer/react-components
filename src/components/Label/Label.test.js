import React from 'react';
import { render, cleanup } from 'react-testing-library';
import 'jest-styled-components';
import Label from '.';

afterEach(cleanup);

describe('Label component', () => {
  it('matches the snapshot', () => {
    const { container } = render(<Label>First Name</Label>);
    expect(container).toMatchSnapshot();
  });

  it('should be rendered with at least one child component', () => {
    const { container } = render(<Label>First Name</Label>);
    expect(container.firstChild.hasChildNodes).toBeTruthy();
  });

  it('should render with default color if not passed explicitely', () => {
    const { container } = render(<Label>First Name</Label>);
    expect(container.firstChild).toHaveStyleRule('color', '#aaa');
  });

  it('should render with float styles if float prop is passed', () => {
    const { container } = render(<Label float>First Name</Label>);
    expect(container.firstChild).toHaveStyleRule('position', 'absolute');
    expect(container.firstChild).toHaveStyleRule('top', '0');
    expect(container.firstChild).toHaveStyleRule('left', '0');
  });
});
