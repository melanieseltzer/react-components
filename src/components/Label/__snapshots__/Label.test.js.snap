// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Label component matches the snapshot 1`] = `
.c0 {
  -webkit-transition: all 0.2s;
  transition: all 0.2s;
  color: #aaa;
}

<div>
  <label
    color="#aaa"
    class="c0"
  >
    First Name
  </label>
</div>
`;
